### PR TITLE
[gpopt] Safer, and more succinct relation cleanup

### PR DESCRIPTION
**Pull request by d**
## TL;DR: Would you like to read and write code like this:

```c++
auto rel = GetRelation(...);
if (!RelIsSupported(rel)) {
	return -1;
}
do_stuff(rel);
```

Instead of code like this:

```c++
Relation rel = GetRelation(...);

if (!RelIsSupported(rel)) {
	CloseRelation(rel);
	return -1;
}

GPOS_TRY {
	do_stuff(rel);
	CloseRelation(rel);
} GPOS_CATCH_EX(ex) {
	CloseRelation(rel);
	GPOS_RETHROW(ex);
} GPOS_CATCH_END;
```
?

## Description
The two snippets above will have the same outcome, but the former is less cluttered and more readable. As a bonus, it's also safer. This patch set introduces a thin wrapper type that encapsulates the resource-safety logic of closing the Postgres Relation object, makes `GetRelation()` return such a type, and cleans up the usages of `GetRelation()`.

### To review
You can either look at the whole diff (but I'd recommend a differ that can strip indentation / whitespace changes), or go through them one by one:
1. first commit cleans up the callers a bit, changing Yoda conditions
1. next commit introduced the wrapper with minimal code to ease the transition
1. Actually transition `GetRleation()` to return the wrapper. Clean up usages.

### Tips for reviewer
1. If you're using GitHub, make sure you teach the UI to ignore whitespace changes like this: 
image

## Feedback I'm seeking:
1. Where should the wrapper class be? I've settled on `gpopt/utils` but I'm not 100% sure it fits there. It just seems to fit there slightly better than all other options I had.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [x] Document changes
- [ ] Communicate in the mailing list if needed
- [x] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp [hunk 1]
```diff
@@ -1261,12 +1214,10 @@ CTranslatorRelcacheToDXL::RetrievePartTableIndex
 	
 	CHAR *index_name = NameStr(index_rel->rd_rel->relname);
 	CMDName *mdname = CDXLUtils::CreateMDNameFromCharArray(mp, index_name);
-	gpdb::CloseRelation(index_rel);
+	index_rel.Close();
```
Comment (chrishajas) [inline src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp:1217]: Why do we need an explicit close here?

File: src/include/gpopt/utils/RelationWrapper.h [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+		return m_relation == nullptr;
+	}
+
+	bool operator!=(std::nullptr_t) const
+	{
+		return m_relation != nullptr;
+	}
+
+	// allows use in typical conditionals
+	explicit operator bool() const
```
Comment (chrishajas) [inline src/include/gpopt/utils/RelationWrapper.h:71]: What does this do?

**Comment by d** [comment]
> Naive question, could you achieve this same cleanup by having gpdb::GetRelation return a unique_ptr and calling gpdb::CloseRelation inside Relation destructor? What are the advantages of RelationWrapper?

Not naïve at all! To make sure we're on the same page, you're envisioning the body of `GetRelation()` to look like this right?

```c++

struct RelationCloser {
    void operator()(Relation r) const {
        gpdb::CloseRelation(r);
    }
}

auto GetRelation(Oid reloid) {
	GP_WRAP_START;
	{
		/* catalog tables: relcache */
		return std::unique_ptr<struct RelationData, RelationCloser>{RelationIdGetRelation(rel_oid)};
	}
	GP_WRAP_END;
}
```

From a client (caller) side perspective, I like this. In fact I have a follow-up PR in mind that removes the implicit conversion operator to `Relation` and instead would force the callers to do a `unique_ptr` style `.get()` to expose the underlying pointer.

From the implementation's PoV, I think the overhead actually reduces readability:

1. Notice the `Deleter` in `unique_ptr` is part of the type. This makes it very cumbersome to use a lambda (as evidenced above)
1. This also makes it slightly more awkward to provide a capability for the client to eagerly close, except for `CloseRelation(rel.get())`;

> gpdb::CloseRelation inside Relation destructor?
Sorry I didn't make it clear, `Relation` is a C-style struct defined in Postgres, so we cannot patch it here.

**Comment by d** [comment]
We had a discussion over Zoom on Friday, for posterity, these are the action items I took:

- [ ] Document why there's an explicit `.Close()` method, and what should a normal use of an instance be
- [ ] Document more clearly the intent of the compatibility layer (`explicit operator bool`)
- [ ] "overload set" is vague and unclear to people who are not intimately familiar with C++, include more plain language to explain the (`=delete`) machinery that protects a programmer from accidental misuse

@chrishajas @ashwinstar did I miss anything from Friday?

**Comment by d** [comment]
I pushed a commit that comments on the rationale behind some design decisions.

**Comment by d** [comment]
@zellerh I'm inviting you for a review because I'm using `auto` here and I know you have a principled distaste for `auto`. I want your blessing here (or a feedback for change on either `auto` or other aspects).

I do have an idea though, if it comes down to changing

```c++
auto rel = GetRelation(...);
```

to

```c++
RelationWrapper rel = GetRelation(...);
```

Then I'm actually inclined to axe the implicit cast support from `RelationWrapper` to `Relation` and make it abundantly explicit as a `.get()` everywhere.

File: src/include/gpopt/utils/RelationWrapper.h [hunk 3]
```diff
@@ -0,0 +1,113 @@
+//---------------------------------------------------------------------------
+//	Greenplum Database
+//	Copyright (C) 2020 VMware, Inc.
+//---------------------------------------------------------------------------
+#ifndef GPDB_RelationWrapper_H
+#define GPDB_RelationWrapper_H
+
+#include <cstddef>
+
+typedef struct RelationData *Relation;
```
Comment (zellerh) [inline src/include/gpopt/utils/RelationWrapper.h:10]: This typedef appears in 3 other places in the code. Is that ok? Shouldn't we have exactly one place where this appears? In general, I'm not a big fan of having a special name for a pointer type, but that's in pre-existing code.

File: src/include/gpopt/utils/RelationWrapper.h (hunk 2 above)
Comment (zellerh) [reply src/include/gpopt/utils/RelationWrapper.h:83]: Do we really need two ways to achieve the same thing, the operator == above and the bool operator here?

File: src/include/gpopt/utils/RelationWrapper.h [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+	/// They are currently unused because the preferred form is
+	///
+	/// \code if (rel) ... \endcode or
+	/// \code if (!rel) ... \endcode
+	bool operator==(std::nullptr_t) const
+	{
+		return m_relation == nullptr;
+	}
+
+	bool operator!=(std::nullptr_t) const
```
Comment (zellerh) [inline src/include/gpopt/utils/RelationWrapper.h:74]: In ORCA, we use NULL everywhere. I object to using nullptr without having at least some discussion.

File: src/include/gpopt/utils/RelationWrapper.h [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+	{
+		return m_relation != nullptr;
+	}
+
+	// behave like a raw pointer on arrow
+	Relation
+	operator->() const
+	{
+		return m_relation;
+	}
```
Comment (zellerh) [inline src/include/gpopt/utils/RelationWrapper.h:93]: Defining these operators can be nice, but I would do this only for classes or groups of classes that are very frequently used, where everyone working in the code will become familiar with this way of overloading operators. When this is used on classes that the average reader will see only rarely, it will just cause confusion instead of helping to make the code easier to read.

File: src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp [hunk 6]
```diff
@@ -1031,9 +998,9 @@ CTranslatorRelcacheToDXL::RetrieveIndex
 {
 	OID index_oid = CMDIdGPDB::CastMdid(mdid_index)->Oid();
 	GPOS_ASSERT(0 != index_oid);
-	Relation index_rel = gpdb::GetRelation(index_oid);
+	auto index_rel = gpdb::GetRelation(index_oid);
```
Comment (zellerh) [inline src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp:1001]: Why not make it explicit that we use RAII, by declaring this as

`RelationWrapper index_rel`

I would find that much easier to read.

File: src/include/gpopt/utils/RelationWrapper.h (hunk 3 above)
Comment (d) [reply src/include/gpopt/utils/RelationWrapper.h:10]: > This typedef appears in 3 other places in the code. Is that ok?

Yes it is OK.

In fact having `Relation` forward declared here is necessary to break a circular dependency with `gpdbwrappers.h`.

Comment (zellerh) [reply src/include/gpopt/utils/RelationWrapper.h:10]: This is not a question of forward declaration, but one of three definitions that could potentially be different.

File: src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp (hunk 6 above)
Comment (d) [reply src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp:1001]: Ah, yes I have a strong opinion here. I think the two sensible design choices we have are:

1. Make `RelationWrapper` a "transparent" wrapper type. By transparent, I mean it behaves just like a `Relation` (except for the proper clean up). In that case, I'm adamant that `auto` is a win for readability, because `GetRelation` already suggests that you're getting a `Relation`-ish object back. And you can perform member access and pass it to functions that expect a `Relation`; or

1. Make `RelationWrapper` an explicit wrapper. This means we remove the nicety of trying to pretend to be a `Relation`. Specifically the implicit cast to `Relation` should be removed. This has the advantage that we don't need to worry about accidentally misusing `CloseRelation`. (OK maybe member access through `operator ->` will still be good ergonomics)

The current patch set takes the first approach. For cleanliness I'd actually favor approach two. The diff footprint will be a bit bigger though, and I have alluded to in comments mentioning "eases migration".

Thoughts?

File: src/include/gpopt/utils/RelationWrapper.h (hunk 3 above)
Comment (d) [reply src/include/gpopt/utils/RelationWrapper.h:10]: I'd expect you to nominate a header `#include` that can replace this `typedef`.

File: src/include/gpopt/utils/RelationWrapper.h (hunk 2 above)
Comment (d) [reply src/include/gpopt/utils/RelationWrapper.h:83]: > Do we really need two ways to achieve the same thing, the operator == above and the bool operator here?

IMO, no we don't. We should prefer to use the `operator bool` here. See my fixup commit that adds comments here.

File: src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp (hunk 6 above)
Comment (zellerh) [reply src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp:1001]: Thanks for the great explanation. Yes, I think I am basically arguing for approach 2, unless we want to use approach 1 as common way of doing things in ORCA.

File: src/include/gpopt/utils/RelationWrapper.h (hunk 5 above)
Comment (d) [reply src/include/gpopt/utils/RelationWrapper.h:93]: > Defining these operators can be nice, but I would do this only for classes or groups of classes that are very frequently used, where everyone working in the code will become familiar with this way of overloading operators. When this is used on classes that the average reader will see only rarely, it will just cause confusion instead of helping to make the code easier to read.

Would you be down to renaming this operator to just:

```
Relation get() const { return m_relation; }
```
?

That'd be more inline with the second approach I outlined in the below comment

File: src/include/gpopt/utils/RelationWrapper.h (hunk 4 above)
Comment (d) [reply src/include/gpopt/utils/RelationWrapper.h:74]: > In ORCA, we use NULL everywhere. I object to using nullptr without having at least some discussion.

I can change the `nullptr` back to `NULL`. How do you feel about `nullptr_t` though? It's the type. Also, if we resolve the review towards making this type less "transparent", then we can axe the two pointer comparison operators altogether.

File: src/include/gpopt/utils/RelationWrapper.h (hunk 3 above)
Comment (zellerh) [reply src/include/gpopt/utils/RelationWrapper.h:10]: Well, this is pre-existing code, so it might be ok. Sorry, if you don't know a good header file to do this, I certainly won't be able to suggest one. Probably this indicates a problem with this approach: We want to do forward-declarations of classes and structs, yet we want to give a pointer to such a class or a struct a special type name. IMHO, that's a bad idea, except for cases that are very common and very obvious for any maintainer of the code. Generally, I expect a pointer to be defined as such. Maybe the Postgres code has a different style, but at least that's the way it is in ORCA.

File: src/include/gpopt/utils/RelationWrapper.h (hunk 5 above)
Comment (zellerh) [reply src/include/gpopt/utils/RelationWrapper.h:93]: Yes, it's just more explicit, and that's the style used in ORCA. That makes overloading the == operator and defining a cast to bool unnecessary. Admittedly a bit more pedestrian, but I find it easier to read, as we generally don't overload operators in ORCA, with rare exceptions.

File: src/include/gpopt/utils/RelationWrapper.h (hunk 4 above)
Comment (zellerh) [reply src/include/gpopt/utils/RelationWrapper.h:74]: Searching ORCA, I found over 11,000 matches for NULL and exactly zero for nullptr.

Comment (d) [reply src/include/gpopt/utils/RelationWrapper.h:74]: > Searching ORCA, I found over 11,000 matches for NULL and exactly zero for nullptr.

11,000 is already fewer than I thought! I'd definitely defer to using a tool like `clang-tidy` to change that for us than introducing an inconsistency here.

BTW, I just did a search, I only found 8495 references to `NULL`, which is _even fewer_ still. Did you include the `NULL` from SQL?

**Comment by ashwinstar** [comment]
No opinions on which way this PR goes, we should still discuss after this PR "we want to use approach 1 as common way of doing things in ORCA or not?" as that didn't seem to get attention as part of this PR.
Once discussed and decision is made for it, would be good to record the same as README to point to developers on coding practice used in project.

**Comment by d** [comment]
I'm working on an alternative by trying to follow the objections above. It's a work-in-progress, but you can preview it over at https://github.com/greenplum-db/gpdb/pull/10743 now. I'll also synthesize all the objections I've gotten so far on this patch, and post it under both PR.

For better posterity, I' rebase and squash commits in this PR now

**Comment by zellerh** [comment]
Thanks a lot! It will be good to see the comments on both of these PRs, then we can decide which way to go.

**Comment by d** [comment]
I summarized the feedback that led to #10743 over there, let me know if I've missed anything! @zellerh

**Comment by d** [comment]
For convenience, this is a verbatim copy of a section of my write-up from #10743. I'm not gonna keep them in sync after posting this:

### Rationale for having two PR's
1. I'm glad that stylistic preferences came up in code review rather than felt later in the team that a change has been "snuck in".
1. I opened #10728 because I cannot stomach the manual clean up in code like this:

   ```c++
   Relation rel = GetRelation(...);

   if (!RelIsSupported(rel)) {
     CloseRelation(rel);
     return -1;
   }

   GPOS_TRY {
     do_stuff(rel);
     int i = calculate_stuff(rel->rd_att->natts);
     CloseRelation(rel);
   } GPOS_CATCH_EX(ex) {
     CloseRelation(rel);
     GPOS_RETHROW(ex);
   } GPOS_CATCH_END;
   ```

   The patch over at #10728 enables us to write code like:

   ```c++
   auto rel = GetRelation(...);
   if (!RelIsSupported(rel)) {
     return -1;
   }
   do_stuff(rel);
   int i = calculate_stuff(rel->rd_att->natts);
   ```

   However, objections were raised in uses of some of the "more C++" idiomatic patterns, such as

   1. `auto rel = GetRelation(...)` makes it unclear what the type of `rel` is, spell out the new type
   1. `operator bool` for a "transparent pointer wrapper" to be used in a conditional like `if (rel)`
   1. `operator ->` for a transparent wrapper to be used for memeber access like `rel->rd_att->natts`
   1. `operator Relation` implicit casts a pointer wrapper to the wrapped pointer type, allowing the caller to directly pass the wrapper to functions that expects a Relation.

   The suggested alternative alternative, described by the reviewer, and I quote verbatim as "pedestrian", results in code like this:

   ```c++
   RelationWrapper rel = GetRelation(...);
   if (!RelIsSupported(rel.get())) {
     return -1;
   }
   do_stuff(rel.get());
   int i = calculate_stuff(rel.get()->rd_att->natts);
   ```

   To be clear, I agree with removing the automatic implicit cast to `Relation` (last objection above). But I disagree with the other points, here are reasons:

   1. > `auto rel = GetRelation(...)` makes it unclear what the type of `rel` is, spell out the new type

      The function name indicates that it returns a `Relation` already. And the variable `behaves like` a Relation in most of its touch points. In fact, not spelling out the type reduces the cognitive load on the reader of the code.

   1. > `operator bool` for a "transparent pointer wrapper" to be used in a conditional like `if (rel)`

      One of the most common thing we do with a returned pointer, at least in ORCA, seems to be a nullity check. This operator enables that. The alternative suggested would be a ["Yoda conditions"](https://en.wikipedia.org/wiki/Yoda_conditions) combined with an explicit access to the wrapped pointer:

      ```c++
      if (NULL == rel.get())
      ```

   1. > `operator ->` for a transparent wrapper to be used for memeber access like `rel->rd_att->natts`

      This is another common use of (formerly) pointers is for member access with arrows. Without this we'd be writing `rel.get()->rd_att->natts`.


   Ultimately though, I made my case and I leave the judgement to the reviewers.

   I care more about _not having to manually clean up 3+ times_ in the code, i.e. I want to push the automatic resource clean up forward, if the resulting code will be uglier because it's littered with `rel.get()->rd_att->natts`, I can hold my nose.

   To be clear, I appreciate the reviewer who pointed out about the inconsistency introduced by of `nullptr`. That reminds me I should really spend time pushing the discussion forward on #10357, which petered out a little because I'm consumed by other priorities.

**Comment by zellerh** [comment]
In my last review I said I would change my comment if it's just me, voicing these concerns. Having heard no others, it seems to me that there is no general objection to changing the programming style and allowing these new constructs, so I retract my -1.
